### ISSUE TITLE
About “Force alignment”

### ISSUE THREAD

**Issue by sunnnnnnnny** [espnet/espnet#2293]
Hello,I found that only English force alignment is supported, right?

**Comment by sw005320** [comment]
It can be used for any language if we use an appropriate pre-trained model.
We actually used it for some Japanese data with the CSJ pre-trained model.

**Comment by sunnnnnnnny** [comment]
I think so too, but I encountered the following two problems:
1、,When I use the pre-trained model of wsj to align the examples provided by git,
when i run 
```
ctc_align_wav.sh --align_dir align   \
   7             --models wsj.transformer.v1    \
   8             --cmvn /data3/test3/zyq/espnet-master/egs/wsj/asr1/align/download/wsj.transformer.v1/data/train_si284/cmvn.ark   \
   9             --align_model /data3/test3/zyq/espnet-master/egs/wsj/asr1/align/download/wsj.transformer.v1/exp/train_si284_pytorch_train_no_preprocess/results/model.acc.       best  \
  10             --align_config conf/decode.yaml \
  11             --dict /data3/test3/zyq/espnet-master/egs/wsj/asr1/align/download/wsj.transformer.v1/data/lang_1char/train_si284_units.txt \
  12             /data3/test3/zyq/espnet-master/egs/wsj/asr1/align/download/wsj.transformer.v1/ctc_align_test.wav "THE SALE OF THE HOTELS IS PART OF HOLIDAY'S STRATEGY TO   SELL OFF ASSETS AND CONCENTRATE ON PROPERTY MANAGEMENT"
```
The decoded text content can be deduplicated and blank to get the same content as the original text label. This is no problem, but the length of the unprocessed result obtained by decoding is not equal to the number of frames of the input voice feature. I am confused.
2、When I run the following code under egs/aishell, the results that appear surprised me：
when i run
```
. ./path.sh
   4
   5 ctc_align_wav.sh --align_dir align \
   6                 --models ais1_v1_model  \
   7                 --cmvn align/download/ais1_v1_model/data/train_sp/cmvn.ark  \
   8                 --align_model  align/download/ais1_v1_model/exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/results/model.last10.avg.best  \
   9                 --align_config conf/decode.yaml  \
  10                 --dict data/lang_1char/train_text.txt \
  11                 BAC009S0767W0122.wav "引领中国房地产广告行业走向新的黄金时代"
```
the result is ：
`<blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 弗 <blank> <blank> <blank> <blank> <blank> <blank> 颇 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 丰 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 图 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 所 <blank> <blank> <blank> <blank> <blank> 圳 <blank> <blank> <blank> <blank> <blank> <blank> 亨 <blank> <blank> <blank> <blank> <blank> <blank> 庄 <blank> <blank> <blank> <blank> <blank> <blank> 呐 <blank> <blank> <blank> <blank> <blank> <blank> 衍 <blank> <blank> <blank> <blank> <blank> <blank> 丛 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 赴 <blank> <blank> <blank> <blank> <blank> 吓 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 方 <blank> <blank> <blank> <blank> 皆 <blank> <blank> <blank> <blank> 黍 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 釜 <blank> <blank> <blank> <blank> <blank> <blank> 旷 <blank> <blank> <blank> <blank> 令 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank>`
The content of the label is completely unmatched with the real text label. Of course, there are cases where the number of input frames and the length of output characters are not equal.


I am learning the principle of forced alignment, so I don’t understand it thoroughly, I hope you can give me some suggestions

**Comment by sw005320** [comment]
> This is no problem, but the length of the unprocessed result obtained by decoding is not equal to the number of frames of the input voice feature. I am confused.

The number of frames after subsampling.
So, it would be 1/4 length from the original one (but subsampling factors depend on the model. Mostly 1/4, but it might be different in some cases).

> 2、When I run the following code under egs/aishell, the results that appear surprised me：
> when i run
> 
> ```
> . ./path.sh
>    4
>    5 ctc_align_wav.sh --align_dir align \
>    6                 --models ais1_v1_model  \
>    7                 --cmvn align/download/ais1_v1_model/data/train_sp/cmvn.ark  \
>    8                 --align_model  align/download/ais1_v1_model/exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/results/model.last10.avg.best  \
>    9                 --align_config conf/decode.yaml  \
>   10                 --dict data/lang_1char/train_text.txt \
>   11                 BAC009S0767W0122.wav "引领中国房地产广告行业走向新的黄金时代"
> ```
> 
> the result is ：
> `<blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 弗 <blank> <blank> <blank> <blank> <blank> <blank> 颇 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 丰 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 图 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 所 <blank> <blank> <blank> <blank> <blank> 圳 <blank> <blank> <blank> <blank> <blank> <blank> 亨 <blank> <blank> <blank> <blank> <blank> <blank> 庄 <blank> <blank> <blank> <blank> <blank> <blank> 呐 <blank> <blank> <blank> <blank> <blank> <blank> 衍 <blank> <blank> <blank> <blank> <blank> <blank> 丛 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 赴 <blank> <blank> <blank> <blank> <blank> 吓 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 方 <blank> <blank> <blank> <blank> 皆 <blank> <blank> <blank> <blank> 黍 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 釜 <blank> <blank> <blank> <blank> <blank> <blank> 旷 <blank> <blank> <blank> <blank> 令 <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank>`
> The content of the label is completely unmatched with the real text label. Of course, there are cases where the number of input frames and the length of output characters are not equal.

Year, this looks very strange.
Can you make sure that the dictionary is correct?
@simpleoier, could you take a look at this case?

**Comment by sunnnnnnnny** [comment]
I also thought it was a dict problem at first，but the dict is constructed by reading the char_list of the downloaded model file model.json( from `align/download/ais1_v1_model/exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/results/model.json`), it should be no problem.

**Comment by simpleoier** [comment]
@sw005320 Thank you for your replying.
@sunnnnnnnny There must be something wrong. I could get exactly the same content as the text to be aligned. Here is my result:
```text
Alignment: <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> 
引 <blank> <blank> <blank> <blank> <blank> <blank> 领 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 
<blank> 中 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 国 <blank> <blank> <blank> <blank> <blank> 
<blank> <blank> 房 <blank> <blank> <blank> <blank> <blank> 地 <blank> <blank> <blank> <blank> <blank> <blank> 产 
<blank> <blank> <blank> <blank> <blank> <blank> 广 <blank> <blank> <blank> <blank> <blank> <blank> 告 <blank> 
<blank> <blank> <blank> <blank> <blank> 行 <blank> <blank> <blank> <blank> <blank> <blank> 业 <blank> <blank> 
<blank> <blank> <blank> <blank> <blank> <blank> <blank> 走 <blank> <blank> <blank> <blank> <blank> 向 <blank> 
<blank> <blank> <blank> <blank> <blank> <blank> <blank> 新 <blank> <blank> <blank> <blank> 的 <blank> <blank> 
<blank> <blank> 黄 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 金 <blank> <blank> <blank> <blank> 
<blank> <blank> 时 <blank> <blank> <blank> <blank>  代 <blank> <blank> <blank> <blank> <blank> <blank> <blank> 
<blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank> <blank>
```

Here is how I got it:
1. I prepared the dictionary by running the ESPnet egs/aishell/asr1 recipe (I ran this about 3 months ago)
2. I downloaded the model from the [egs/aishell/asr1](https://github.com/espnet/espnet/blob/master/egs/aishell/asr1/RESULTS.md), model link is [here](https://drive.google.com/open?id=1BIQBpLRRy3XSMT5IRxnLcgLMirGzu8dg).
3. The following are exact commands (I am using the latest ESPnet)
```shell
cd egs/aishell/asr1
. ./path.sh

download_from_google_drive.sh https://drive.google.com/open\?id\=1BIQBpLRRy3XSMT5IRxnLcgLMirGzu8dg download/aishell.transformer.v1 tar.gz 
text=$(grep BAC009S0767W0122 /my/path/to/aishell/asr1/data/test/text | awk '{print $2}')
ctc_align_wav.sh \
  --align_dir align \
  --cmvn download/aishell.transformer.v1/data/train_sp/cmvn.ark \
  --align-model download/aishell.transformer.v1/exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/results/model.last10.avg.best \
  --align-config download/aishell.transformer.v1/conf/decode.yaml \
  --dict /my/path/to/aishell/asr1/data/lang_1char/train_sp_units.txt \
/my/path/to/aishell/asr1/data/data_aishell/wav/test/S0767/BAC009S0767W0122.wav ${text}
```

**Comment by sunnnnnnnny** [comment]
I found that the char_list that I got from `pretrain_model/exp/train_sp_pytorch_train_pytorch_transformer_lr1.0/results/model.json` is wrong as a dict. I regenerated the dict and everything works fine now,Thank you for your meticulous guidance.
